File: algorithms/leetcode/rust/src/algos/p1_two_sum.rs
use std::collections::HashMap;

pub struct Solution;

// fn main() {
//   let v = vec![1, 2, 3, 4];
//   let b = Solution::two_sum(v, 4);
//   println!("{:#?}", b);
// }

// [3,2,4], 6
impl Solution {
  pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut m: HashMap<i32, usize> = HashMap::new();

    for (i, n) in nums.iter().enumerate() {
      if let Some(&c) = m.get(n) {
        if i == c {
          continue;
        }

        return vec![i as i32, c as i32];
      } else {
        let c = target - *n;
        m.insert(c, i);
      }
    }

    vec![]
  }
}

File: algorithms/leetcode/python/p1_two_sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
      num_idxs = {}

      for i in range(len(nums)):
        num_idxs[nums[i]] = i

      print(num_idxs)

      for i in range(len(nums)):
        comp = target - nums[i]

        if comp in num_idxs:
            idx = num_idxs[comp]
            if idx == i:
              continue
            return [idx, i]
      return []

solution = Solution()
print(solution.twoSum([3,3], 6))
